File: item-list/src/main/java/com/gurunars/item_list/MutatedUpdateFetcher.java
package com.gurunars.item_list;

import java.util.ArrayList;
import java.util.List;

class MutatedUpdateFetcher<ItemType extends Item> {

    List<? extends Change<ItemType>> get(List<ItemType> sourceMiddle,
                                         List<ItemType> targetMiddle,
                                         int startOffset) {
        List<Change<ItemType>> changes = new ArrayList<>();

        for (ItemType sourceItem : sourceMiddle) {
            int index = targetMiddle.indexOf(sourceItem);
            int realIndex = startOffset + index;
            ItemType targetItem = targetMiddle.get(index);
            if (!sourceItem.payloadsEqual(targetItem)) {
                changes.add(new ChangeUpdate<>(targetItem, realIndex, realIndex));
            }
        }

        return changes;
    }

}

File: item-list/src/main/java/com/gurunars/item_list/MutatedUpdateFetcher.kt
package com.gurunars.item_list

import java.util.ArrayList

internal class MutatedUpdateFetcher<ItemType : Item> {

    operator fun get(sourceMiddle: List<ItemType>,
                     targetMiddle: List<ItemType>,
                     startOffset: Int): List<Change<ItemType>> {
        val changes = ArrayList<Change<ItemType>>()

        for (sourceItem in sourceMiddle) {
            val index = targetMiddle.indexOf(sourceItem)
            val realIndex = startOffset + index
            val targetItem = targetMiddle[index]
            if (!sourceItem.payloadsEqual(targetItem)) {
                changes.add(ChangeUpdate(targetItem, realIndex, realIndex))
            }
        }

        return changes
    }

}
